smarts/env/gymnasium/utils/action_conversion.py

```diff
@@ -0,0 +1,254 @@
+            dtype=np.float64,
+        )
+        for _ in range(MPC_ARRAY_COUNT)
+    ]
+    + [
+        gym.spaces.Box(
+            low=np.array([DT_MINIMUM] * TRAJECTORY_LENGTH),
+            high=np.array([DT_MAXIMUM] * TRAJECTORY_LENGTH),
+            dtype=np.float64,
+        )
+    ]
+)
+
+
+@dataclass(frozen=True)
+class _FormattingGroup:
+    space: gym.Space
+    formatting_func: Callable[[Any], Any] = field(default=_DEFAULT_PASSTHROUGH)
+
+
+@lru_cache(maxsize=1)
+def get_formats() -> Dict[ActionSpaceType, _FormattingGroup]:
+    """Returns the formatting groups for each of the avaliable action types."""
+    return {
+        ActionSpaceType.ActuatorDynamic: _FormattingGroup(
+            space=_actuator_dynamic_space,
+        ),
+        ActionSpaceType.Continuous: _FormattingGroup(
+            space=_continuous_space,
+        ),
+        ActionSpaceType.Direct: _FormattingGroup(
+            space=_direct_space,
+        ),
+        ActionSpaceType.Empty: _FormattingGroup(
+            space=gym.spaces.Tuple(spaces=()),
+            formatting_func=lambda a: None,
+        ),
+        ActionSpaceType.Lane: _FormattingGroup(
+            space=_lane_space,
+            formatting_func=_format_lane_space,
+        ),
+        ActionSpaceType.LaneWithContinuousSpeed: _FormattingGroup(
+            space=_lane_with_continuous_speed_space,
+        ),
+        ActionSpaceType.MPC: _FormattingGroup(
+            space=_mpc_space,
+        ),
+        ActionSpaceType.MultiTargetPose: _FormattingGroup(
+            space=_multi_target_pose_space,
+        ),
+        ActionSpaceType.RelativeTargetPose: _FormattingGroup(
+            space=_relative_target_pose_space,
+        ),
+        ActionSpaceType.TargetPose: _FormattingGroup(
+            space=_target_pose_space,
+        ),
+        ActionSpaceType.Trajectory: _FormattingGroup(
+            space=_trajectory_space,
+        ),
+        ActionSpaceType.TrajectoryWithTime: _FormattingGroup(
+            space=_trajectory_with_time_space,
+        ),
+    }
+
+
+class ActionOptions(IntEnum):
+    """Defines the options for how the formatting matches the action space."""
+
+    multi_agent = 0
+    """Action must map to partial action space. Only active agents are included."""
+    full = 1
+    """Action must map to full action space. Inactive and active agents are included."""
+    unformatted = 2
+    """Actions are not reformatted or constrained to action space. Actions must directly map to
+    underlying SMARTS actions."""
+    default = 0
+    """Defaults to :attr:`multi_agent`."""
+
+
+class ActionSpacesFormatter:
+    """Formats actions to adapt SMARTS to `gym` environment requirements."""
+
+    def __init__(
+        self, agent_interfaces: Dict[str, AgentInterface], action_options: ActionOptions
+    ) -> None:
+        self._agent_interfaces = agent_interfaces
+        self._action_options = action_options
```

Comment:
+ Consider explaining inputs (agent_interfaces, action_options) in the docstring.
+ Helpful VSCode extension for generating docstrings:
    https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring